File: 12_Advanced_Graphs/03_Network_Delay_Time/0743-network-delay-time.cpp
/*
Problem: LeetCode 743 - Network Delay Time

Description:
You are given a network of n nodes, labeled from 1 to n. You are also given times, a list of travel times
as directed edges times[i] = (ui, vi, wi), where ui is the source node, vi is the target node,
and wi is the time it takes for a signal to travel from source to target.
We will send a signal from a given node k. Return the time it takes for all the n nodes to receive
the signal. If it is impossible for all the n nodes to receive the signal, return -1.

Intuition:
This problem can be solved using Dijkstra's algorithm, which finds the shortest paths
from a source node to all other nodes in a weighted graph.

Approach:
1. Create a graph representation using an adjacency list, where graph[u] contains a list
   of pairs (v, w), representing an edge from node u to node v with weight w.
2. Use Dijkstra's algorithm to find the shortest path from the source node k to all other
   nodes in the graph.
3. Initialize a min heap (priority queue) to keep track of the nodes to visit, and set the
   distance of the source node k to 0 and distances of all other nodes to infinity.
4. Push the source node k into the min heap.
5. While the min heap is not empty, pop the node with the minimum distance from the heap.
6. For each neighbor of the current node, update the distance if a shorter path is found
   and push the neighbor into the min heap if it hasn't been visited yet.
7. After processing all nodes, the distance array will contain the shortest time to reach
   each node from the source node k. Return the maximum value in the distance array, which
   represents the time it takes for all nodes to receive the signal.
8. If there are nodes that cannot be reached from the source node k, the distance array
   will have some nodes with infinite distance. In this case, return -1.

Time Complexity:
The time complexity of Dijkstra's algorithm is O(E*logV), where E is the number of edges
and V is the number of vertices.

Space Complexity:
The space complexity is O(V+E), where V is the number of vertices and E is the number of edges,
due to the adjacency list representation of the graph.
*/

class Solution {
  public:
    int networkDelayTime(vector<vector<int>> &times, int n, int k) {
        // Create a graph representation using an adjacency list
        unordered_map<int, vector<pair<int, int>>> graph;

        for (const auto &time : times) {
            graph[time[0]].push_back({time[1], time[2]});
        }

        // Initialize a min heap (priority queue) to track nodes to visit
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        vector<int> distance(n + 1, INT_MAX);
        // Set distance of source node k to 0 and push it into the min heap
        distance[k] = 0;
        pq.push({0, k});

        while (!pq.empty()) {
            auto [dist, node] = pq.top();
            pq.pop();

            // Check if the node has been visited already
            if (dist > distance[node]) {
                continue;
            }

            // Update the distances for neighbors
            for (const auto &neighbor : graph[node]) {
                int nextNode = neighbor.first;
                int newDist = dist + neighbor.second;

                if (newDist < distance[nextNode]) {
                    distance[nextNode] = newDist;
                    pq.push({newDist, nextNode});
                }
            }
        }

        // Find the maximum distance, which represents the time for all nodes to receive the signal
        int maxDist = 0;

        for (int i = 1; i <= n; ++i) {
            maxDist = max(maxDist, distance[i]);
        }

        // If some nodes cannot be reached, return -1
        return maxDist == INT_MAX ? -1 : maxDist;
    }
};

File: 12_Advanced_Graphs/03_Network_Delay_Time/0743-network-delay-time.rs
/*
Problem: LeetCode 743 - Network Delay Time

Key Idea:
The key idea is to find the minimum time it takes for a signal to reach all nodes in a weighted directed graph. This problem can be solved using Dijkstra's algorithm, a greedy algorithm for finding the shortest path in a weighted graph.

Approach:
1. Create a vector "dist" to store the minimum time to reach each node from the source node (initialized with infinity for all nodes except the source).
2. Create a priority queue (min heap) to keep track of the next node to visit based on their estimated distances.
3. Start with the source node, set its distance to 0, and push it into the priority queue.
4. While the priority queue is not empty:
   - Pop the node with the smallest estimated distance from the priority queue.
   - If this node has been visited before, skip it.
   - For each neighbor of the current node:
     - Calculate the tentative distance to the neighbor through the current node.
     - If the tentative distance is smaller than the current distance stored in "dist," update the distance and push the neighbor into the priority queue.
5. After processing all nodes, check if any node has infinite distance. If so, it's not reachable, so return -1. Otherwise, return the maximum distance in "dist" as the answer.

Time Complexity:
O(N^2 + E * log E), where N is the number of nodes and E is the number of edges. In the worst case, Dijkstra's algorithm can have O(N^2) complexity if using a naive priority queue. Using a binary heap-based priority queue can optimize it to O(E * log E).

Space Complexity:
O(N + E), where N is the number of nodes for the "dist" vector, and E is the number of edges for the adjacency list.
*/

use std::collections::{BinaryHeap, HashMap};

impl Solution {
    pub fn network_delay_time(times: Vec<Vec<i32>>, n: i32, k: i32) -> i32 {
        // Create an adjacency list to represent the graph.
        let mut graph: HashMap<i32, Vec<(i32, i32)>> = HashMap::new();
        for time in &times {
            let (u, v, w) = (time[0], time[1], time[2]);
            graph.entry(u).or_insert(Vec::new()).push((v, w));
        }

        // Initialize distances with infinity, except for the source.
        let mut dist: Vec<i32> = vec![std::i32::MAX; n as usize];
        dist[k as usize - 1] = 0;

        // Create a min heap to store (distance, node) pairs.
        let mut min_heap: BinaryHeap<(i32, i32)> = BinaryHeap::new();
        min_heap.push((0, k));

        while let Some((d, u)) = min_heap.pop() {
            if d > dist[u as usize - 1] {
                continue;
            }
            if let Some(neighbors) = graph.get(&u) {
                for (v, w) in neighbors {
                    let new_dist = dist[u as usize - 1] + w;
                    if new_dist < dist[*v as usize - 1] {
                        dist[*v as usize - 1] = new_dist;
                        min_heap.push((new_dist, *v));
                    }
                }
            }
        }

        // Check if any node is unreachable.
        if dist.iter().any(|&d| d == std::i32::MAX) {
            return -1;
        }

        // Return the maximum distance in "dist."
        *dist.iter().max().unwrap()
    }
}
